qiskit/transpiler/passes/mapping/algorithm/ancestors.py

```diff
@@ -0,0 +1,71 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Utility for speeding up a function to find ancestors in DAG.
+"""
+
+from functools import lru_cache
+
+import networkx as nx
+
+CACHE_SIZE = 2 ** 10
+
+
+class Ancestors:
+    """
+    Utility class for speeding up a function to find ancestors in DAG.
+    """
+
+    def __init__(self, G: nx.DiGraph, max_depth=5):
+        self._graph = G.reverse()
+        self._max_depth = max_depth
+        self._depth = 0
+
+    def ancestors(self, n: int) -> set:
+        """
+        Ancestor nodes of the node `n`
+        Args:
+            n: Index of the node in the `self._graph`
+        Returns:
+            Set of indices of the ancestor nodes.
+        """
+        self._depth = 0
```

Comment:
This seems like a complicated mix between an iterative and recursive approach, where you modify an instance attribute to track how deep into a recursive call you are when executing one of the methods. This `depth` should instead be an argument to `_ancestors_rec`, where the default is `0`, and you pass `depth + 1` to the recursive call.

Reply:
You are right. Because it is complicated, I will simplify it. Thanks.

Reply:
I simplified it in https://github.com/Qiskit/qiskit-terra/pull/1803/commits/504c8a6f30508ab8603ae2862a3203aca09b4414.